Sources/OpenPass/DeviceAuthorizationFlow.swift

```diff
@@ -0,0 +1,211 @@
+/// The flow then supports polling to check for authorization.
+///
+///     Task {
+///        let flow = OpenPassManager.shared.deviceAuthorizationFlow
+///        do {
+///            // Request a Device Code
+///            let deviceCode = try await flow.fetchDeviceCode()
+///
+///            // Present the code and uri to the user
+///
+///            // Poll for authorization
+///            let tokens = try await flow.fetchAccessTokenPolling(deviceCode: deviceCode)
+///            // do something with tokens
+///        } catch {
+///            // do something with error
+///        }
+///    }
+@MainActor
+public final class DeviceAuthorizationFlow {
+
+    // When polling the token endpoint, it's possible that the response could ask us to "slow down". 
+    // For each slow down response we must add an additional 5 seconds to the original interval.
+    internal var slowDownMultiplier: Int64 = 0
+
+    // The number of additional seconds that should be added to the interval if asked to slow down (the polling).
+    private static let defaultSlowDownFactor: Int64 = 5
+
+    // MARK: - Init
+    
+    private let openPassClient: OpenPassClient
+    private let tokenValidator: IDTokenValidation
+    private let tokensObserver: ((OpenPassTokens) async -> Void)
+    private let dateGenerator: DateGenerator
+    private let clock: Clock
+
+    internal init(
+        openPassClient: OpenPassClient,
+        tokenValidator: IDTokenValidation,
+        tokensObserver: @escaping ((OpenPassTokens) async -> Void),
+        dateGenerator: DateGenerator = .init { Date() },
+        clock: Clock = RealClock()
+    ) {
+        self.openPassClient = openPassClient
+        self.tokenValidator = tokenValidator
+        self.tokensObserver = tokensObserver
+        self.dateGenerator = dateGenerator
+        self.clock = clock
+    }
+
+    // MARK: - Public API
+    
+    /// Start the authorization flow by requesting a Device Code from the API server.
+    /// The returned ``DeviceCode`` is also used with the `fetchAccessToken` and`fetchAccessTokenPolling` methods.
+    /// - Returns: A Device Code representation
+    public func fetchDeviceCode() async throws -> DeviceCode {
+        // Reset in case the flow is reused
+        slowDownMultiplier = 0
+
+        let authorizeDeviceCodeResponse = try await openPassClient.getDeviceCode()
+        switch authorizeDeviceCodeResponse {
+        case .success(let response):
+            return DeviceCode(response: response, now: dateGenerator.now)
+        case .failure(let error):
+            throw OpenPassError.unableToGenerateDeviceCode(name: error.error, description: error.errorDescription)
+        }
+    }
+
+    /// Fetch an access token, if authorized.
+    /// If authorization is still pending, this method returns `nil`
+    /// See also `fetchAccessTokenPolling`.
+    public func fetchAccessToken(deviceCode: DeviceCode) async throws -> OpenPassTokens? {
```

Comment:
Nit: Is this a public function, and therefore part of our API, that isn't used in production code? Do we expect (or want) consumers to poll the endpoint themselves? I'm curious what value this has for us to expose, and whether we should just remove to limit our API.